xarray/core/variable.py

```diff
@@ -1632,6 +1632,73 @@ def rolling_window(self, dim, window, window_dim, center=False,
             array, axis=self.get_axis_num(dim), window=window,
             center=center, fill_value=fill_value))
 
+    def coarsen(self, windows, func, side='left', trim_excess=False):
+        windows = {k: v for k, v in windows.items() if k in self.dims}
+        new_dimensions = {k: utils.get_temp_dimname(self.dims, k)
+                          for k in windows}
+        reshaped = self._coarsen_reshape(windows, side, trim_excess,
+                                         new_dimensions)
+
+        axis = tuple([reshaped.get_axis_num(d) for d
+                      in new_dimensions.values()])
+        return type(self)(self.dims, func(reshaped, axis=axis), self._attrs)
+
+    def _coarsen_reshape(self, windows, side, trim_excess, coarsen_dimensions):
+        """
+        Construct a reshaped-variable for corsen
+        """
+        if not utils.is_dict_like(side):
+            side = {d: side for d in windows.keys()}
+
+        if not utils.is_dict_like(trim_excess):
+            trim_excess = {d: trim_excess for d in windows.keys()}
+
+        # remove unrelated dimensions
+        side = {k: v for k, v in side.items() if k in self.dims}
+        trim_excess = {k: v for k, v in trim_excess.items() if k in self.dims}
+
+        if windows == {}:
+            return type(self)(self.dims, self.data, self._attrs)
+
+        for d, window in windows.items():
+            if window <= 0:
+                raise ValueError('window must be > 0')
+
+        variable = self
+        for d, window in windows.items():
+            # trim or pad the object
+            size = variable.shape[self._get_axis_num(d)]
+            n = int(size / window)
+            if trim_excess[d]:
+                if side[d] == 'left':
+                    variable = variable.isel({d: slice(window * int(n))})
+                else:
+                    excess = size - window * n
+                    variable = variable.isel({d: slice(excess, None)})
+            else:  # pad
+                pad = window * (n + 1) - size
+                if side[d] == 'left':
+                    pad_widths = {d: (0, pad)}
+                else:
+                    pad_widths = {d: (pad, 0)}
+                variable = variable.pad_with_fill_value(pad_widths)
+
+        shape = []
+        axes = []
+        dims = []
+        for i, d in enumerate(variable.dims):
+            if d in windows:
+                size = variable.shape[i]
+                shape.append(int(size / windows[d]))
+                shape.append(windows[d])
+                dims.append(d)
+                dims.append(coarsen_dimensions[d])
+            else:
+                shape.append(variable.shape[i])
+                dims.append(d)
+
+        return Variable(dims, variable.data.reshape(shape), self._attrs)
```

Comment:
Is it worth making an actual xarray.Variable object here rather than just returning `data`? With the need to come up with new dimension names, I think it might be cleaner avoiding that.